File: cco13p5.cpp
#include <bits/stdc++.h>
using namespace std;


struct point {
	int x, y;
	point operator - (const point& rhs) const { return point{ x - rhs.x, y - rhs.y }; }
	double operator * (const point& rhs) const { return x * rhs.x + y * rhs.y; }
	double cross (const point& rhs) const { return x * rhs.y - y * rhs.x; }
	double mag() const { return sqrt(x * x + y * y); }
};

bool equal(double a, double b, double eps = 1e-5) {
	return abs(a - b) < eps;
}


void doit() {
	int n;
	cin >> n;

	vector<point> v1(n), v2(n);
	for (auto& [x, y] : v1)
		cin >> x >> y;
	for (auto& [x, y] : v2)
		cin >> x >> y;
	

	auto encode = [](point p1, point p2, point p3) -> double {
		point a = p1 - p2;
		point b = p3 - p2;

		double
			dot = a * b,
			cross = a.cross(b),
			magA = a.mag(),
			magB = b.mag();
		
		// double angle = acos(dot / (magA * magB));
		double angle = atan2(cross, dot);
		double ratio = magA / magB;

		return angle * ratio;
	};
	
	vector<double> s(n), t(2*n);
	for (int i = 0; i < n; ++i) {
		s[i]          = encode(v1[i], v1[(i+1)%n], v1[(i+2)%n]);
		t[i] = t[i+n] = encode(v2[i], v2[(i+1)%n], v2[(i+2)%n]); // mod
	}


	auto kmp = [](const vector<double>& s, const vector<double>& t) -> int {
		int n = s.size(),
			m = t.size();

		vector<int> lps(n);
		for (int i = 1; i < n; ++i) {
			int j = lps[i-1];
			while (j > 0 && !equal(s[i], s[j]))
				j = lps[j-1];
			if (equal(s[i], s[j]))
				++j;
			
			lps[i] = j;
		}

		for (int i = 0, j = 0; i < m; ) {
			if (equal(s[j], t[i]))
				++i, ++j;
			
			if (j == n) {
				return i - j; // one
				j = lps[j - 1];
			}
			else if (i < m && !equal(s[j], t[i])) {
				if (j) {
					j = lps[j-1];
				}
				else {
					++i;
				}
			}
		}
		return -1;
	};

	cout << (kmp(s, t) + 1) << "\n";
}


int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	int t;
	cin >> t;
	for (int i = 0; i < t; ++i)
		doit();
}

File: cco13p5.rs
#[macro_use] extern crate dmoj;

// #[derive(Debug)]
struct Point {
    x: i32,
    y: i32,
}
fn dot  (lhs: &Point, rhs: &Point) -> f32 { (lhs.x * rhs.x + lhs.y * rhs.y) as f32 }
fn cross(lhs: &Point, rhs: &Point) -> f32 { (lhs.x * rhs.y - lhs.y * rhs.x) as f32 }
fn mag  (lhs: &Point) -> f32 { ((lhs.x * lhs.x + lhs.y * lhs.y) as f32).sqrt() }

const EPSILON: f32 = 1e-4;
fn equal(a: f32, b: f32) -> bool { (a - b).abs() < EPSILON }

fn encode(p1: &Point, p2: &Point, p3: &Point) -> f32 {
	let a: Point = Point{
		x: p1.x - p2.x,
		y: p1.y - p2.y,
	};
	let b: Point = Point{
		x: p3.x - p2.x,
		y: p3.y - p2.y,
	};

	let mag_a: f32 = mag(&a);
	let mag_b: f32 = mag(&b);
	let d_prod: f32 = dot(&a, &b);
	let x_prod: f32 = cross(&a, &b);

	let angle: f32 = f32::atan2(d_prod, x_prod);
	let ratio: f32 = mag_a / mag_b;

	angle * ratio
}

fn kmp(s: &Vec<f32>, t: &Vec<f32>) -> i32 {
	let n: usize = s.len();
	let m: usize = t.len();

	let mut lps: Vec<usize> = vec![Default::default(); n];
	for i in 1..n {
		let mut j: usize = lps[i-1];
		while j > 0 && !equal(s[i], s[j]) {
			j = lps[j-1];
		}
		if equal(s[i], s[j]) {
			j += 1;
		}
		lps[i] = j;
	}

	let mut i: usize = 0;
	let mut j: usize = 0;
	while i < m {
		if equal(s[j], t[i]) {
			i += 1;
			j += 1;
		}
		if j == n {
			return (i - j) as i32;
			// j = lps[j-1];
		}
		else if i < m && !equal(s[j], t[i]) {
			if j != 0 {
				j = lps[j-1];
			}
			else {
				i += 1;
			}
		}
	}
	-1
}

fn scan_points(n: usize) -> Vec<Point> {
	(0..n)
		.map(|_| Point{
			x: scan!(i32),
			y: scan!(i32),
		})
		.collect()
}

fn solve() {
	let n: usize = scan!(usize);
	
	let v1: Vec<Point> = scan_points(n);
	let v2: Vec<Point> = scan_points(n);
	
	let s: Vec<f32> = (0..n)
		.map(|i| encode(&v1[i], &v1[(i+1)%n], &v1[(i+2)%n]))
		.collect();
	let mut t: Vec<f32> = (0..n)
		.map(|i| encode(&v2[i], &v2[(i+1)%n], &v2[(i+2)%n]))
		.collect();
	t.append(&mut t.clone());
	
	println!("{}", kmp(&s, &t)+1);
}

fn main() {
	for _ in 0..scan!(usize) {
		solve();
	}
}
